### ISSUE TITLE
memory-mapped faiss indices

### ISSUE THREAD

**Issue by alexanderguzhva** [facebookresearch/faiss#4101]
# Summary

A prototype of memory-mapped files that contain Faiss Indices (IndexFlatCodes and HNSW at this moment).

https://github.com/zilliztech/knowhere/pull/996

@mdouze Could you please give your comments on the approach, just in case I've missed some important corner cases? Despite the link is from the Faiss fork, it should be pretty straightforward to understand the changes. Basically, I replace `std::vector<>` with a custom container that supports mmap-ping.

The PR for Faiss will be implemented as soon as we agree on the details.
Thanks.

**Comment by mdouze** [comment]
Yes that seems very good. 
I started a similar effort but never got a chance to finalize it, see . So please go ahead if you want to commit it. 
For ref [here is a patch]([shipit-facebookresearch-faiss-D56685773.patch](https://gist.github.com/mdouze/735481b7c4aed552dbddb1b7aba85522#file-shipit-facebookresearch-faiss-d56685773-patch)) of what I was working on.

**Comment by navneet1v** [comment]
@alexanderguzhva this is very promising if we can replace the `std::vector<>` with the container. I think beyond mmap it has a lot of usecases. are you working on it? If not I can pick this up. 

@mdouze what are your thoughts on how we should move forward in the contribution.

**Comment by alexanderguzhva** [comment]
@navneet1v I will publish a PR next week, I'm waiting for some extra things to be verified, such as using data from some 3-rd party preallocated memory area as well

**Comment by navneet1v** [comment]
> [@navneet1v](https://github.com/navneet1v) I will publish a PR next week, I'm waiting for some extra things to be verified, such as using data from some 3-rd party preallocated memory area as well

thank you for the update. Excited to see this feature coming out.

**Comment by alexanderguzhva** [comment]
#4199

**Comment by satymish** [comment]
Hi @alexanderguzhva, should we close this issue?

**Comment by alexanderguzhva** [comment]
@satymish , no, because #4199 and #4212 were commited and immediately reverted in https://github.com/facebookresearch/faiss/commit/fbc7db2cce762a86676094c24c7fb639d7c608a2 and https://github.com/facebookresearch/faiss/commit/15491a1e4f5a513a8684e5b7262ef4ec22eda19d because of Meta's internal problems with unit tests. This is being worked on
